Write-only argument GA (#36572)
hashicorp/terraform: website/docs/language/resources/ephemeral/write-only.mdx

Unlike other ephemeral constructs in Terraform, such as ephemeral resources or variables, write-only arguments also accept both ephemeral and non-ephemeral values.

## Requirements

To use write-only arguments, you must use Terraform v.1.11 or later and use  a `resource` that supports write-only arguments.

## Declare a write-only argument

Providers indicate in the Terraform registry whether an argument is write-only. For example, the `aws` provider's `aws_db_instance` resource has a write-only `password_wo` argument. The `password_wo` argument accepts a value to use as the database password:

<CodeBlockConfig highlight="7">

```hcl
resource "aws_db_instance" "test" {
  instance_class      = "db.t5.micro"
  allocated_storage   = "5"
  engine              = "postgres"
  username            = "admin"
  skip_final_snapshot = true
  password_wo = <ephemeral or non-ephemeral value>
}
```

</CodeBlockConfig>

> Write-only arguments accept both ephemeral and non-ephemeral values. For example, you could also use a string as the value of a write-only argument:

Reviewer: Maybe we should say somewhere this is not recommended though. 

Should we add password version to these examples as well for completeness?

Author: Yeah, good point about adding the versions - I like that for setting a nice precedent. 

How would it feel to make that a positive reframe to instead focus on recommending v.s. advising against? Not for any particular reason other than I know folks always like to see a recommendation: 
```
Write-only arguments accept both ephemeral and non-ephemeral values. For example, you could also use a string as the value of a write-only argument:

<example>

However, we recommend using write-only arguments for passing ephemeral values to resources.
```

Reviewer: ya that sounds great! I think we should provide an opinionated view in the docs

```hcl
resource "aws_db_instance" "test" {
  instance_class      = "db.t5.micro"
  allocated_storage   = "5"
  engine              = "postgres"
  username            = "admin"

>   skip_final_snapshot = true
>   password_wo = "my-password-here"

Reviewer 2: ```suggestion
  skip_final_snapshot = true
  password_wo         = "my-password-here"
```

```hcl
resource "aws_db_instance" "test" {
  instance_class      = "db.t5.micro"
  allocated_storage   = "5"
  engine              = "postgres"
  username            = "admin"
  skip_final_snapshot = true
  password_wo         = "new-password-here"
  password_wo_version = 1
}
```

</CodeBlockConfig>

The provider now uses the write-only argument value when creating the `aws_db_instance` resource and when the `password_wo_version` version argument increments. The `password_wo_version` version argument is stored in state, allowing Terraform to track when the write-only argument changes.

To trigger an update of a write-only argument, increment the version argument's value in your configuration:

```hcl

resource "aws_db_instance" "main" {
  instance_class      = "db.t5.micro"
  allocated_storage   = "5"
  engine              = "postgres"
  username            = "admin"
  password_wo         = "new-password-here"
  password_wo_version = 2
}
```

When you increment the `password_wo_version` argument, Terraform notices that change in its plan and notifies the `aws` provider. The `aws` provider then uses the new `password_wo` value to update the `aws_db_instance` resource.

> <!--- TODO: add Vault example once the provider supports write-only args --->

Author: I [used the example from the blog](https://docs.google.com/document/d/1aVO8QqoQWUgrgIzreEFp2MCraMZFnOLn-qizMcUr_8U/edit?usp=sharing) for this article, and plan to update once we have a Vault example ready to go, but if anyone has another more sophisticated example we could add I would super appreciate it!

Reviewer: Vault should be working on this in march (hopefully)

  allocated_storage   = "5"
  engine              = "postgres"
  username            = "admin"
  skip_final_snapshot = true
  password_wo         = <ephemeral or non-ephemeral value>
  password_wo_version = 1
}
```

</CodeBlockConfig>

Write-only arguments accept both ephemeral and non-ephemeral values. For example, you could also use a string as the value of a write-only argument:

```hcl
resource "aws_db_instance" "test" {
  instance_class      = "db.t5.micro"
  allocated_storage   = "5"
  engine              = "postgres"
  username            = "admin"
  skip_final_snapshot = true
  password_wo         = "my-password-here"
  password_wo_version = 1
}
```

However, we recommend using write-only arguments for passing ephemeral values to resources. For example, you can use an `ephemeral` resource to generate a random password and pass it to the `password_wo` write-only argument:

<CodeBlockConfig highlight="11">

```hcl
ephemeral "random" "password" {
  length = 16
}

resource "aws_db_instance" "test" {
  instance_class      = "db.t5.micro"
  allocated_storage   = "5"
  engine              = "postgres"
  username            = "admin"
  skip_final_snapshot = true

>   password_wo          = ephemeral.random.password.value

Author: ```suggestion
  password_wo         = ephemeral.random.password.value
```

> resource "aws_db_instance" "test" {
>   instance_class      = "db.t5.micro"
>   allocated_storage   = "5"
>   engine              = "postgres"
>   username            = "admin"
>   skip_final_snapshot = true
>   password_wo = ephemeral.random.password.value
>   password_wo_version = 1
> }

Reviewer 2: ```suggestion
resource "aws_db_instance" "test" {
  instance_class      = "db.t5.micro"
  allocated_storage   = "5"
  engine              = "postgres"
  username            = "admin"
  skip_final_snapshot = true
  password_wo = ephemeral.random.password.value
  password_wo_version = 1
}
```
```suggestion
resource "aws_db_instance" "test" {
  instance_class      = "db.t5.micro"
  allocated_storage   = "5"
  engine              = "postgres"
  username            = "admin"
  skip_final_snapshot = true
  password_wo          = ephemeral.random.password.value
  password_wo_version = 1
}
```

Terraform also cannot create plan diffs for write-only arguments because it does not store them in plan files, so providers only use write-only arguments when creating a new resource.

Providers implement version arguments to let practitioners track write-only argument values and control when a provider uses those write-only arguments. The implementation of write-only arguments and their version arguments is provider-specific, so consult the Registry for more details about your specific provider.

For example, the `aws_db_instance` resource has an accompanying `password_wo_version` argument for the `password_wo` write-only argument:

<CodeBlockConfig highlight="11-12">

```hcl
resource "aws_db_instance" "test" {
  instance_class      = "db.t5.micro"
  allocated_storage   = "5"
  engine              = "postgres"
  username            = "admin"
  skip_final_snapshot = true
  password_wo         = "new-password-here"
  password_wo_version = 1
}
```

</CodeBlockConfig>

> The provider now uses the write-only argument value when creating the `aws_db_instance` resource and when the `password_wo_version` version argument increments. The `password_wo_version` version argument is stored in state, allowing Terraform to track when the write-only argument changes.

Reviewer 2: ```suggestion
The provider now uses the write-only argument value when creating the `aws_db_instance` resource and whenever you increment the `password_wo_version` version argumen. The `password_wo_version` version argument is stored in state, allowing Terraform to track when the write-only argument changes.
```

---
page_title: Use write-only arguments
description: Learn how to use write-only arguments to set temporary values that are not stored in Terraform's state or plan files.
---

# Use write-only arguments

> Write-only arguments let you securely pass temporary values to Terraform resources during a Terraform operation without persisting those values to state or plan files. Use write-only arguments to store sensitive data for the current Terraform operation, such as a short-lived password, token, or session identifier. 

Reviewer 2: ```suggestion
Write-only arguments let you securely pass ephemeral values to Terraform resources during a Terraform operation without persisting those values to state or plan files. Use write-only arguments to store sensitive data for the current Terraform operation, such as a password, token, or session identifier. 
```

## Background

Write-only arguments are only available during the runtime of a Terraform operation, and Terraform omits those values from state and plan files.

> **Hands-on**: Declare a write-only argument in the [Upgrade RDS major version](/terraform/tutorials/aws/rds-upgrade) tutorial.

> You can use write-only arguments to pass ephemeral values directly to a resource without Terraform persisting that value. For example, you can directly pass a generated password or a fetched token to a write-only argument without risking Terraform exposing that value. Use write-only arguments to ensure sensitive data only exists during the current Terraform operation.

Reviewer 2: This might be re-iterating what the intro says a bit too much? Maybe the intro should be more focused on the "why" and this para more on the "how"?

Author: Lol thank you - you can tell I rewrote both of these way too many times 😂

I kept the intro the same because I do feel like it covers the "why" pretty well, but expanded the background to explain a bit more of how you can use ephemeral values _together_. Which I think is the context that I was missing here, and what fits best in a background section

> Unlike other ephemeral constructs in Terraform, such as ephemeral resources or variables, write-only arguments also accept both ephemeral and non-ephemeral values.

Reviewer 2: ```suggestion
Unlike other ephemeral constructs in Terraform, such as ephemeral resources or variables, write-only arguments accept both ephemeral and non-ephemeral values.
```

To use write-only arguments, you must use Terraform v.1.11 or later and use a resource that supports write-only arguments.

## Declare a write-only argument

Providers indicate in the Terraform registry whether an argument is write-only. For example, the `aws` provider's `aws_db_instance` resource has a write-only `password_wo` argument. The `password_wo` argument accepts a value to use as the database password:

<CodeBlockConfig highlight="7">

```hcl
resource "aws_db_instance" "test" {
  instance_class      = "db.t5.micro"
  allocated_storage   = "5"
  engine              = "postgres"
  username            = "admin"
  skip_final_snapshot = true
  password_wo         = <ephemeral or non-ephemeral value>
  password_wo_version = 1
}
```

</CodeBlockConfig>

Write-only arguments accept both ephemeral and non-ephemeral values. For example, you could also use a string as the value of a write-only argument:

```hcl
resource "aws_db_instance" "test" {
  instance_class      = "db.t5.micro"
  allocated_storage   = "5"
  engine              = "postgres"
  username            = "admin"
  skip_final_snapshot = true
  password_wo         = "my-password-here"
  password_wo_version = 1
}
```

> However, we recommend using write-only arguments for passing ephemeral values to resources. For example, you can use an `ephemeral` resource to generate a random password and pass it to the `password_wo` write-only argument:

Reviewer 2: ```suggestion
However, we recommend using write-only arguments for passing ephemeral values to resources. If you use a non-ephemeral value, that value will either be stored in your Terraform state, or hard-coded in your configuration. For example, you can use an `ephemeral` resource to generate a random password and pass it to the `password_wo` write-only argument:
```

Trying to get to the "why" we recommend, but ignore if too much.

<CodeBlockConfig highlight="11">

```hcl
ephemeral "random" "password" {
  length = 16
}

resource "aws_db_instance" "test" {
  instance_class      = "db.t5.micro"
  allocated_storage   = "5"
  engine              = "postgres"
  username            = "admin"
  skip_final_snapshot = true
  password_wo         = ephemeral.random.password.value
  password_wo_version = 1
}
```

</CodeBlockConfig>

During a Terraform operation, the provider uses the `password_wo` value to configure the database instance, and then Terraform discards that value without storing it in the plan or state file. 

## Update write-only arguments with versions

Terraform does not store write-only arguments in state files, so Terraform has no way of knowing if a write-only argument value has changed. Because Terraform cannot track write-only argument values, it sends write-only arguments to the provider during every operation.

> Terraform also cannot create plan diffs for write-only arguments because it does not store them in plan files, so providers only use write-only arguments when creating a new resource.

Reviewer 3: I think this sentence implies that providers cannot use write-only arguments to update a resource. 
```suggestion
Terraform also cannot create plan diffs for write-only arguments because it does not store them in plan files, so providers typically implement mechanisms such as version arguments to indicate a plan diff to Terraform.
```

Author: I was trying to explain what happens if you _don't_ add an accompanying version arg (because I know most of them are optional), but I like your suggestion of going straight to version args. How does the following feel for making your suggestion a bit shorter? Since the next paragraph explains the why behind version args:
```suggestion
Terraform also cannot create plan diffs for write-only arguments because it does not store those values in plan files. However, providers typically include version arguments alongside write-only arguments. Terraform stores version arguments in state, and can track if a version argument changes. 
```

Reviewer 3: That sounds good to me!

Providers implement version arguments to let practitioners track write-only argument values and control when a provider uses those write-only arguments. The implementation of write-only arguments and their version arguments is provider-specific, so consult the Registry for more details about your specific provider.

For example, the `aws_db_instance` resource has an accompanying `password_wo_version` argument for the `password_wo` write-only argument:

<CodeBlockConfig highlight="11-12">

```hcl
resource "aws_db_instance" "test" {
  instance_class      = "db.t5.micro"
  allocated_storage   = "5"
  engine              = "postgres"
  username            = "admin"
  skip_final_snapshot = true

>   password_wo         = "new-password-here"

Author: ```suggestion
  password_wo         = "old-password-here"
```

  password_wo_version = 1
}
```

</CodeBlockConfig>

> The provider now uses the write-only argument value when creating the `aws_db_instance` resource and whenever you increment the `password_wo_version` version argument. The `password_wo_version` version argument is stored in state, allowing Terraform to track when the write-only argument changes.

Author: ```suggestion
The provider uses the write-only argument value when creating the `aws_db_instance` resource and stores the `password_wo_version` version argument in state.
```
